Improved code generation patterns for 2022.3

tspence opened the pull request:
Fixed spacing issues in helptext

tspence committed:
Merge pull request #1 from Lockstep-Network/main

Merge latest from origin

Commit:
Latest code generation pattern with updated helptext

Commit:
Revised with Manish's code generator patterns

tspence commented:
Using feedback from Manish and Rishi, I've updated the Ruby code generator.  Changes include:

* All file names, class names, and data model names are fully snake cased 
* Added attr_accessors to the lockstep_api class
* Updated the "request" method based on Manish's fixes
* Within each API call, defined the "params" object using field names
* Fixed an issue that caused path parameters to be ignored in the function definition
* Fixed an issue where the Ruby keyword "include" was used as a variable name, which was not allowed

rishirajkumar97 commented:
Just a clarification , are we sticking to using Structs rather than use Classes  for each of the model ? @tspence

rishirajkumar97 commented on lib/lockstep_sdk/clients/activities.rb:
```diff
@@ -0,0 +1,110 @@
+    # 
+    # @param id [uuid] The unique Lockstep Platform ID number of this Activity
+    # @param include_param [string] To fetch additional data on this object, specify the list of elements to retrieve. Available collections: Company, Attachments, CustomFields, Notes, References, and UserAssignedToName
+    def retrieve_activity(id:, include_param:)
```
think instead of `include_param` we could just  rename it to `params`, here and all places should work I guess

tspence replied:
The API parameter is called "include", but I can't use the name "include" here because that's a Ruby keyword.  I wanted to add something to the name to make it not a keyword - would there be something better?

tspence commented:
> Just a clarification , are we sticking to using Structs rather than use Classes for each of the model ? @tspence

I'm open for what you might suggest!  What is the best practice today?  Should we switch to classes?

Commit:
Latest suggestions from Rishi

Commit:
Updated data models with snake case

rishirajkumar97 commented on lib/lockstep_sdk/models/data_models.rb:
```diff
@@ -21,110 +21,110 @@ module LockstepSdk
     # the priority of the task, and any amounts collected, paid, or credited for the task.
     class activity_model
```
Camel Casing for class Names is required

tspence replied:
As per discussion on Slack, we're switching to ProperCase (aka PascalCase)

rishirajkumar97 commented on lib/lockstep_sdk/models/data_models.rb:
```diff
@@ -21,110 +21,110 @@ module LockstepSdk
         # @return [activity_xref_model] All references attached to this applied activity. To retrieve this collection, specify `References` in the "Include" parameter for your query.
-        attr_accessor :references            
+        attr_accessor :references
     )
```
this ending brackets would be replaced with the keyword `end`

tspence replied:
Fixed, thanks!

Commit:
Updated class definitions in Ruby as per Rishi

Commit:
Updated gemspec with reference to "awrence"
Changed naming for data models to classes, using ProperCase
Updated type hints with correct names
Removed duplicate LockstepApi class definitions

ManishNarayan commented:
Hey @tspence, Writing down some of the issues I noticed here - 

1. In the lockstep_api.rb file I had made some changes which are not reflected here. 
    a. The Authentication check in the request method has a mistake where we are checking the api_key both the times.
    b. I would recommend swapping the lines in with_api_key method so that the user gets the correct feedback.
Please update this file with the one from my PR. You can check if these changes make sense.

2. I see that each model class is inside has a LockstepSdk module. Do we need this? This would mean some extra code for the  user to instantiate the model. Unless we have plans to add another class into the module, just a class should suffice.

3. I have added as_json and to_json methods for each model class in my PR. This was required since we would need to pass a json object of the model when we call create API. So add those too.

4. The filenames of models and clients could also be changed. Clients would be plurals, ex : activities.rb  and  Models would be singular, ex : activity.rb. The class names inside would be the same but in ProperCase (aka PascalCase) like you said.

5. We would need to change the version to  0.1.0 ( mainly in the version.rb file). This is the standard according to what I've read and we should follow that. We'll have to change that in all the files which have the version in comments.

6. We don't need the old clients and DataModels.rb file which are outside the lockstep_sdk directory. If we go ahead with the clients and models using classes, the old files can be removed.

These are my suggestions @tspence but I hope even @rishirajkumar97  can look at the comments and let me know if I've gone wrong somewhere! 

All these changes are in my PR, so we can just take a look at that code if anything required!

ManishNarayan commented:
Another point I noticed was that,  when we call create APIs the platform does not expect all the attributes to be passed. 

An ex - https://developer.lockstep.io/reference/post_api-v1-activities. When this activity is created we don't have to pass like - @created, @created_user_id, @modified, @modified_user_id,  @amount_collected and even the attachments, notes etc. 

If we include these in the initialize method, ruby will expect them to be passed while creating the model. I had removed them from my model files.

Apologies for not mentioning this before when I raised a PR. Just wanted to raise the issue now so that, we can have a discussion on this during the call later.

tspence commented:
> Hey @tspence, Writing down some of the issues I noticed here -
> 
> 1. In the lockstep_api.rb file I had made some changes which are not reflected here.
>    a. The Authentication check in the request method has a mistake where we are checking the api_key both the times.
>    b. I would recommend swapping the lines in with_api_key method so that the user gets the correct feedback.
>    Please update this file with the one from my PR. You can check if these changes make sense.
> 2. I see that each model class is inside has a LockstepSdk module. Do we need this? This would mean some extra code for the  user to instantiate the model. Unless we have plans to add another class into the module, just a class should suffice.
> 3. I have added as_json and to_json methods for each model class in my PR. This was required since we would need to pass a json object of the model when we call create API. So add those too.
> 4. The filenames of models and clients could also be changed. Clients would be plurals, ex : activities.rb  and  Models would be singular, ex : activity.rb. The class names inside would be the same but in ProperCase (aka PascalCase) like you said.
> 5. We would need to change the version to  0.1.0 ( mainly in the version.rb file). This is the standard according to what I've read and we should follow that. We'll have to change that in all the files which have the version in comments.
> 6. We don't need the old clients and DataModels.rb file which are outside the lockstep_sdk directory. If we go ahead with the clients and models using classes, the old files can be removed.
> 
> These are my suggestions @tspence but I hope even @rishirajkumar97 can look at the comments and let me know if I've gone wrong somewhere!
> 
> All these changes are in my PR, so we can just take a look at that code if anything required!

Thanks so much for these detailed feedback items!

For your items:
1. I agree - I will doublecheck your PR and make sure to pull in all your changes into the code generator.
2. The purpose of grouping the client classes together is to reduce the number of functions on a single class.  Our API is already relatively large and we have dozens of endpoints.  Using the "client.activity.create_activity()" pattern means that autocomplete gives you a shorter list of options at each stage during typing.  I've followed that pattern in the other four SDKs but if we feel strongly that it's not appropriate for Ruby we can reconsider.
3. I will add the to_json and as_json methods from your examples to the code generator pattern, thank you!
4. The filenames of the clients and models are driven by the Swagger file.  We do transform them by changing their case, so we can convert them to PascalCase, camelCase, or snake_case if we want; but otherwise the text names should be taken from swagger.
5. The version number is driven by the API version number.  The goal is to match the SDK's version number to the version of the API when it was built.  You can see the API version number here: https://api.sbx.lockstep.io/
6. Good point, I will make sure to delete those.

Regarding your other question - about the initializer methods - should we make all these initializer parameters optional?

Commit:
Merge branch 'lockstep-main'

# Conflicts:
#	lib/clients/Activities.rb
#	lib/clients/ApiKeys.rb
#	lib/clients/AppEnrollments.rb
#	lib/clients/Applications.rb
#	lib/clients/Attachments.rb
#	lib/clients/CodeDefinitions.rb
#	lib/clients/Companies.rb
#	lib/clients/Contacts.rb
#	lib/clients/CreditMemoApplied.rb
#	lib/clients/Currencies.rb
#	lib/clients/CustomFieldDefinitions.rb
#	lib/clients/CustomFieldValues.rb
#	lib/clients/Definitions.rb
#	lib/clients/Emails.rb
#	lib/clients/InvoiceHistory.rb
#	lib/clients/Invoices.rb
#	lib/clients/Leads.rb
#	lib/clients/Notes.rb
#	lib/clients/PaymentApplications.rb
#	lib/clients/Payments.rb
#	lib/clients/Provisioning.rb
#	lib/clients/Reports.rb
#	lib/clients/Status.rb
#	lib/clients/Sync.rb
#	lib/clients/UserAccounts.rb
#	lib/clients/UserRoles.rb
#	lib/lockstep_sdk/clients/activities.rb
#	lib/lockstep_sdk/clients/api_keys.rb
#	lib/lockstep_sdk/clients/app_enrollments.rb
#	lib/lockstep_sdk/clients/applications.rb
#	lib/lockstep_sdk/clients/attachments.rb
#	lib/lockstep_sdk/clients/code_definitions.rb
#	lib/lockstep_sdk/clients/companies.rb
#	lib/lockstep_sdk/clients/contacts.rb
#	lib/lockstep_sdk/clients/credit_memo_applied.rb
#	lib/lockstep_sdk/clients/currencies.rb
#	lib/lockstep_sdk/clients/custom_field_definitions.rb
#	lib/lockstep_sdk/clients/custom_field_values.rb
#	lib/lockstep_sdk/clients/definitions.rb
#	lib/lockstep_sdk/clients/emails.rb
#	lib/lockstep_sdk/clients/invoice_history.rb
#	lib/lockstep_sdk/clients/invoices.rb
#	lib/lockstep_sdk/clients/leads.rb
#	lib/lockstep_sdk/clients/notes.rb
#	lib/lockstep_sdk/clients/payment_applications.rb
#	lib/lockstep_sdk/clients/payments.rb
#	lib/lockstep_sdk/clients/provisioning.rb
#	lib/lockstep_sdk/clients/reports.rb
#	lib/lockstep_sdk/clients/status.rb
#	lib/lockstep_sdk/clients/sync.rb
#	lib/lockstep_sdk/clients/user_accounts.rb
#	lib/lockstep_sdk/clients/user_roles.rb
#	lib/lockstep_sdk/lockstep_api.rb
#	lib/lockstep_sdk/version.rb
#	lib/models/DataModels.rb
#	lockstepsdk.gemspec

Commit:
Generate clients for 2022.3

tspence commented:
Changes since previous PR:
* Renamed all classes using PascalCase and endpoint clients are now PascalCaseClient
* Fixed camelback json of the body
* Corrected use of keyword "include"
* All the models now have as_json and to_json functions
* Redefined the initializer for models so they can accept null values

Commit:
Remove required awrence for client classes

ManishNarayan commented:
LGTM!
